src/epoch/participant.rs

```diff
@@ -0,0 +1,176 @@
+// Manages a single participant in the epoch scheme. This is where all
+// of the actual epoch management logic happens!
+
+use std::{fmt, mem};
+use std::cell::UnsafeCell;
+use std::sync::atomic::{self, AtomicUsize, AtomicBool};
+
+use epoch::{Atomic, Guard, garbage, global};
+use epoch::participants::ParticipantNode;
+
+/// The garbage collection threshold (in garbage items).
+const GC_THRESHOLD: usize = 128;
+
+/// Thread-local data for epoch participation.
+pub struct Participant {
+    /// The local epoch.
+    ///
+    /// This is incremented on new epochs.
+    epoch: AtomicUsize,
+    /// Number of pending uses of `epoch::pin()`.
+    ///
+    /// Keeping a count allows for reentrant use of epoch management.
+    in_critical: AtomicUsize,
+    /// The thread-local garbage tracking.
+    garbage: UnsafeCell<garbage::Local>,
+    /// Is the thread still active?
+    ///
+    /// This becomes `false` when the thread exits. This is ultimately used to free `Participant`
+    /// records.
+    pub active: AtomicBool,
+    /// The next participant node.
+    ///
+    /// The participant list is coded intrusively; here's the pointer to the next node.
+    pub next: Atomic<ParticipantNode>,
+}
+
+impl Participant {
+    /// Enter a critical section.
+    ///
+    /// This method is reentrant, allowing for nested critical sections.
+    ///
+    /// It returns `true` is this is the first entry on the stack (as opposed to a re-entrant
+    /// call).
+    pub fn enter(&self) -> bool {
+        // Increment the counter.
+        if self.in_critical.fetch_add(1, atomic::Ordering::Relaxed) > 0 {
```

Comment:
I expect this would make pinning more expensive. On Intel, you've essentially replaced a load->store->mfence sequence with a lock xadd -> mfence sequence, which isn't too different than a mfence->mfence sequence. Maybe the second fence is fast and can be ordered around since the previous fence already blocked everything, but it seems inefficient at first glance. The story is better but not great on ll/sc architectures. Since it's not a big change in clarity over load->add->store I think it should get changed back.